src/main/java/com/emc/pravega/perf/WriterWorker.java

```diff
@@ -55,33 +57,79 @@
-            String key;
-            if (isRandomKey) {
-                key = Integer.toString(workerID + rand.nextInt());
-            } else {
-                key = Integer.toString(workerID);
+        perf.benchmark();
+        return null;
+    }
+
+    private class eventsWriter implements performance {
+
+        public void benchmark() throws InterruptedException, ExecutionException {
+            CompletableFuture retFuture = null;
+            Random rand = new Random();
+
+            for (int i = 0; i < events; i++) {
+
+                // Construct event payload
+                String val = System.currentTimeMillis() + ", " + workerID + ", " + (int) (Math.random() * 200);
+                String payload = String.format("%-" + messageSize + "s", val);
+                String key;
+                if (isRandomKey) {
+                    key = Integer.toString(workerID + rand.nextInt());
+                } else {
+                    key = Integer.toString(workerID);
+                }
+
+                final Instant startTime = Instant.now();
+                retFuture = writeData(key, payload);
+                // event ingestion
+                retFuture = stats.recordTime(retFuture, startTime, payload.length());
             }
 
-            final Instant startTime = Instant.now();
-            retFuture = writeData(key, payload);
-            // event ingestion
-            retFuture = stats.recordTime(retFuture, startTime, payload.length());
+            flush();
+
+            //Wait for the last packet to get acked
+            retFuture.get();
         }
+    }
 
-        flush();
+    private class eventspersecWriter implements performance {
 
-        //Wait for the last packet to get acked
-        retFuture.get();
+        public void benchmark() throws InterruptedException, ExecutionException {
+            CompletableFuture retFuture = null;
+            Random rand = new Random();
 
-        return null;
+            while (Duration.between(StartTime, Instant.now()).getSeconds() < secondsToRun) {
+                final Instant loopStart = Instant.now();
+                long loopTime;
+
+                for (int i = 0; ((loopTime = Duration.between(loopStart, Instant.now()).toMillis()) < 1000) &&
+                        (i < events); i++) {
+
+                    // Construct event payload
+                    String val = System.currentTimeMillis() + ", " + workerID + ", " + (int) (Math.random() * 200);
+                    String payload = String.format("%-" + messageSize + "s", val);
+                    String key;
+                    if (isRandomKey) {
+                        key = Integer.toString(workerID + rand.nextInt());
+                    } else {
+                        key = Integer.toString(workerID);
+                    }
+
+                    final Instant startTime = Instant.now();
+                    retFuture = writeData(key, payload);
```

Comment:
As I understand, the purpose of this PR is to create a "rate limiter" of the events written in the system. While I could have missed something, I don't see anywhere in the code where are we preventing writers to write as fast as they can (I would expect a "sleep" or something similar to distribute writes along time based on some input rate).

Reply:
after this loop, we sleep for 1000-looptime milliseconds

Reply:
I think we need a sleep after "each" write, otherwise you are generating bursts of requests that are not keeping a constant rate of events per second. And this will impact the latency of operations.